server/src/domain/storage/storage.core.ts

```diff
@@ -143,26 +168,93 @@ export class StorageCore {
+          );
+          return;
+        }
+      }
 
       move = await this.moveRepository.update({ id: move.id, oldPath: actualPath, newPath });
     } else {
       move = await this.moveRepository.create({ entityId, pathType, oldPath, newPath });
     }
 
+    if (pathType === AssetPathType.ORIGINAL && !assetInfo) {
+      this.logger.warn(`Unable to complete move. Missing asset info for ${entityId}`);
+      return;
+    }
+
     if (move.oldPath !== newPath) {
-      await this.repository.moveFile(move.oldPath, newPath);
+      try {
+        this.logger.debug(`Attempting to rename file: ${move.oldPath} => ${newPath}`);
+        await this.repository.rename(move.oldPath, newPath);
+      } catch (err: any) {
+        if (err.code !== 'EXDEV') {
+          this.logger.warn(
+            `Unable to complete move. Error renaming file with code ${err.code} and message: ${err.message}`,
+          );
+          return;
+        }
+        this.logger.debug(`Unable to rename file. Falling back to copy, verify and delete`);
+        await this.repository.copyFile(move.oldPath, newPath);
+
+        if (!(await this.verifyNewPathContentsMatchesExpected(move.oldPath, newPath, assetInfo))) {
+          this.logger.warn(`Skipping move due to file size mismatch`);
+          return;
+        }
+
+        try {
+          await this.repository.unlink(move.oldPath);
+        } catch (err: any) {
+          this.logger.warn(`Unable to delete old file, it will now no longer be tracked by Immich: ${err.message}`);
+        }
+      }
     }
+
     await this.savePath(pathType, entityId, newPath);
     await this.moveRepository.delete(move);
   }
 
+  private async verifyNewPathContentsMatchesExpected(
+    oldPath: string,
+    newPath: string,
+    assetInfo?: { sizeInBytes: number; checksum: Buffer },
+  ) {
+    const oldPathSize = assetInfo ? assetInfo.sizeInBytes : (await this.repository.stat(oldPath)).size;
+    const newPathSize = (await this.repository.stat(newPath)).size;
+    this.logger.debug(`File size check: ${newPathSize} === ${oldPathSize}`);
+    if (newPathSize !== oldPathSize) {
+      this.logger.warn(`Unable to complete move. File size mismatch: ${newPathSize} !== ${oldPathSize}`);
+      return false;
+    }
+    if (assetInfo && (await this.configCore.getConfig()).storageTemplate.hashVerificationEnabled) {
+      const { checksum } = assetInfo;
+      const newChecksum = await this.cryptoRepository.hashFile(newPath);
+      if (!newChecksum.equals(checksum)) {
+        this.logger.warn(
+          `Unable to complete move. File checksum mismatch: ${newChecksum.toString('base64')} !== ${checksum.toString(
+            'base64',
+          )}`,
+        );
+        return false;
+      }
+      this.logger.debug(`File checksum check: ${newChecksum.toString('base64')} === ${checksum.toString('base64')}`);
+    }
+    return true;
+  }
+
```

Comment:
This will always fail in cases where users have changed the file on-disk behind Immich's back (but maybe that's OK, just good to be aware of)

Reply:
That's not something we support and they shouldn't be doing it, so I think failing here is fine.